src/PrestaShopBundle/Form/Admin/Sell/Product/Pricing/SpecificPriceType.php

```diff
@@ -0,0 +1,306 @@
+use Symfony\Component\Validator\Constraints\GreaterThanOrEqual;
+use Symfony\Component\Validator\Constraints\NotBlank;
+use Symfony\Component\Validator\Constraints\Type;
+
+class SpecificPriceType extends TranslatorAwareType
+{
+    /**
+     * @var string
+     */
+    private $defaultCurrencyIso;
+
+    /**
+     * @var FormChoiceProviderInterface
+     */
+    private $currencyByIdChoiceProvider;
+
+    /**
+     * @var FormChoiceProviderInterface
+     */
+    private $countryByIdChoiceProvider;
+
+    /**
+     * @var FormChoiceProviderInterface
+     */
+    private $groupByIdChoiceProvider;
+
+    /**
+     * @var FormChoiceProviderInterface
+     */
+    private $shopByIdChoiceProvider;
+
+    /**
+     * @var FormChoiceProviderInterface
+     */
+    private $taxInclusionChoiceProvider;
+
+    /**
+     * @var ConfigurableFormChoiceProviderInterface
+     */
+    private $combinationIdChoiceProvider;
+
+    /**
+     * @var UrlGeneratorInterface
+     */
+    private $urlGenerator;
+
+    /**
+     * @param TranslatorInterface $translator
+     * @param array $locales
+     * @param string $defaultCurrencyIso
+     * @param FormChoiceProviderInterface $currencyByIdChoiceProvider
+     * @param FormChoiceProviderInterface $countryByIdChoiceProvider
+     * @param FormChoiceProviderInterface $groupByIdChoiceProvider
+     * @param FormChoiceProviderInterface $shopByIdChoiceProvider
+     * @param FormChoiceProviderInterface $taxInclusionChoiceProvider
+     * @param ConfigurableFormChoiceProviderInterface $configurableFormChoiceProvider
+     * @param UrlGeneratorInterface $urlGenerator
+     */
+    public function __construct(
+        TranslatorInterface $translator,
+        array $locales,
+        string $defaultCurrencyIso,
+        FormChoiceProviderInterface $currencyByIdChoiceProvider,
+        FormChoiceProviderInterface $countryByIdChoiceProvider,
+        FormChoiceProviderInterface $groupByIdChoiceProvider,
+        FormChoiceProviderInterface $shopByIdChoiceProvider,
+        FormChoiceProviderInterface $taxInclusionChoiceProvider,
+        ConfigurableFormChoiceProviderInterface $configurableFormChoiceProvider,
+        UrlGeneratorInterface $urlGenerator
+    ) {
+        parent::__construct($translator, $locales);
+        $this->currencyByIdChoiceProvider = $currencyByIdChoiceProvider;
+        $this->countryByIdChoiceProvider = $countryByIdChoiceProvider;
+        $this->groupByIdChoiceProvider = $groupByIdChoiceProvider;
+        $this->shopByIdChoiceProvider = $shopByIdChoiceProvider;
+        $this->taxInclusionChoiceProvider = $taxInclusionChoiceProvider;
+        $this->defaultCurrencyIso = $defaultCurrencyIso;
+        $this->combinationIdChoiceProvider = $configurableFormChoiceProvider;
+        $this->urlGenerator = $urlGenerator;
+    }
+
+    public function buildForm(FormBuilderInterface $builder, array $options): void
+    {
+        if (!isset($builder->getData()['product_id'])) {
```

Comment:
tried it at first if I remember correctly. But Its more convenient to have it as data (because **only data** travels to FormDataHandler and I can set **only data** in FormDataProvider).. So in options it would be kinda redundant as i would still only be using data.

Reply:
Well actually in the FormBuilder you can define a `FormOptionsProviderInterface` in your service definition which is able to create an array of options based on the data (this is used for product form only for now)

Although it may be overkill and it also has limitations for now (especially you can't define options for sub-levels form types right now... only the first layer)

So this one is ok for now